token/core/tms.go

```diff
@@ -67,39 +67,22 @@ func (m *TMSProvider) GetTokenManagerService(opts driver.ServiceOptions) (servic
 		return nil, errors.Errorf("namespace not specified")
 	}
 	m.lock.Lock()
-	invokeCallback := false
-	defer func() {
-		// unlock
-		m.lock.Unlock()
-		// invoke callback
-		if invokeCallback && m.callbackFunc != nil {
-			err = m.callbackFunc(service, opts.Network, opts.Channel, opts.Namespace)
-			if err != nil {
-				logger.Fatalf("failed to initialize tms for [%s]: [%s]", opts, err)
-			}
-		}
-	}()
+	defer m.lock.Unlock()
 
 	key := tmsKey(opts)
```

Comment:
We can avoid most locks in such cases with double-checked locking:
```
key := tmsKey(opts)
if service, ok := m.services[key]; ok {
 return service, nil
}
m.lock.Lock()
defer m.lock.Unlock()
if service, ok := m.services[key]; ok {
 return service, nil
}
service, err := m.getTokenManagerService(opts)
if err != nil {
 return nil, err
}
m.services[key] = service
return service, nil
```

Reply:
👍